### ISSUE TITLE
covariance matrix as well as correlation matrix

### ISSUE THREAD

**Issue by thistleknot** [atoti/atoti#407]

**Comment by APolyakova** [comment]
Hi @thistleknot how are you? 
Could you kindly elaborate on this desired feature, is it in the context of portfolio aggregation?

**Comment by thistleknot** [comment]
Yes it is. Covariance matrix is a matrix that is derived from a matrix.
Standard aggregate features seem to aggregate only on columns

**Comment by APolyakova** [comment]
Hi @thistleknot if you wish, pls provide a calculation example describing what you want?

**Comment by thistleknot** [comment]
Notebook:
https://github.com/thistleknot/Python-Stock/blob/master/code/markowitz/Markowitz-slim-Atoti.ipynb

Function

def random_portfolio(TREASURY_BILL_RATE,returns):

    w = rand_weights(len(pd.DataFrame(returns).columns))

    p = np.asmatrix(np.mean(returns, axis=0))

    #print(returns)
    C = np.array(pd.DataFrame(returns).cov().values)
    #print(C)

    mu = np.asmatrix(w) * p.T
    #print(w)
    sigma = np.sqrt(np.asmatrix(w) * C * np.asmatrix(w).T)

    sharpe = (mu-TREASURY_BILL_RATE)/sigma
    #sharpe = mu / sigma

    return np.float(mu), np.float(sigma), np.float(sharpe),
np.array(w, dtype=object)


The way sigma is derived.  Well, basically that whole calculation.  As
far as I can tell I can't do anything with a matrix in atoti.  Just a
column at a time.  The way you're team modelled sigma there initially
is they did it via column (that is a type of sigma, but not for this
calculation which calculates the covariance amongst a set of
predictors, which is essential for modern portfolio theory).

**Comment by thistleknot** [comment]
Anastasia,

https://nbviewer.org/github/thistleknot/Python-Stock/blob/master/code/markowitz/Markowitz-slim-Atoti.ipynb

In this part of my notebook (I meant to send this in nbviewer last time, so
I apologize for that), you can see where I was about to make use of atoti
in commented lines, and I got stopped because I basically can't access the
above function at that junction (lol, that's punny).

I was going to take a training query, and then feed that into a markowitz
function (i.e. the last email).  But I'm stopped because I can't do the
matrix calculation as an aggregate option.  I figured at that point, I
might as well just not use atoti at all.

#m["Lookback Window"] = 252
n_portfolio = 100
stocksp = 0.2mfundsp = 0.1bondsp = 0.2commp = 0.2
nran = 200nstocks = 20
holdout_returns = []holdout_returns_weighted = []
#simulation
randomSet = list(np.sort(np.random.choice(range(0, len(Training)), nran)))
for tensor in randomSet:

    bestSharpe = 0
    runs = 0
    weight = None
    bestMean = 0
    bestStd = 0
    means = np.array([])
    stds = np.array([])
    sharpes = np.array([])
    bestWeights = None
    weights = np.array([])

    randomsymbols = list(np.random.choice(pf.data.columns, nstocks,
replace=False))

    Training_dates = list(Training[tensor].ravel())

    TREASURY_BILL_RATE =
FRED_pvt.interpolate(method="time")[FRED_pvt.index==Training_dates[-1]][interest_free_measure].dropna().values[-1]
    print(TREASURY_BILL_RATE)

    Holdout_dates = list(Holdout[tensor].ravel())

    #this is where atoti currently lacks.  I can't derive a covariance
matrix without a full array (see random_portfolio function).
    Training_query = returns.loc[Training_dates][randomsymbols]

    Holdout_query = returns.loc[Holdout_dates][randomsymbols]

    '''    Training_query = \    (        cube.query(
m["value.MEAN"],            levels=[l["variable"]],
condition=(            (l["variable"].isin(*randomsymbols))
& (l["Date"].isin(*Training_dates))            ),        )    )    '''

    names = Training_query.columns

    '''    Holdout_query = \    (        cube.query(
m["value.MEAN"],            levels=[l["variable"]],
condition=(            (l["variable"].isin(*randomsymbols))
& (l["Date"].isin(*Holdout_dates))            ),        )    )    '''

    #display([dates[0],dates[-1]])
    #display(query.T)

    #Monte Carlo efficient frontier
    for _ in range(n_portfolio):
        means2, stds2, sharpes2, weights2 = np.column_stack(
            [random_portfolio((TREASURY_BILL_RATE*1/252.75),Training_query)
for _ in range(len(names))]
        )
        means_ = means2
        stds_ = stds2
        sharpes_ = sharpes2
        weights_ = weights2

        bestWeights = weights_[sharpes_.argmax()]

        means = np.append(means, means_)
        stds = np.append(stds, stds_)
        sharpes = np.append(sharpes, sharpes_)
        weights = np.append(weights, weights_)

    df_ = pd.DataFrame(weights[sharpes.argmax()]).set_index(names)
    df_.columns = ["Weights"]

    #next4weeks = returns.loc[[i[0] for i in Holdout[tensor]]][randomsymbols]

    #df_.sort_values(by=["Weights"], ascending=False, inplace=True)

    display(df_.T)

    wr = np.array(
        weighted_return(
            np.array((np.exp(pd.DataFrame(Holdout_query.mean()).T) -
1).cumsum().tail(1)),
            weights[sharpes.argmax()],
        )
    )[0][0]

    max_sr_ret = means[sharpes.argmax()]
    max_sr_vol = stds[sharpes.argmax()]
    max_sharpe = sharpes[sharpes.argmax()]

    print("non weighted return Training")
    display(Training_query.mean().mean())
    print("non weighted Return Holdout")
    display(pd.DataFrame(Holdout_query.mean()).T)
    #display(pd.DataFrame(Holdout_query.mean()).T.mean())

    print("optimal Weights & Return holdout")
    display(pd.DataFrame(bestWeights).set_index(names).T)
    display(wr)

    #display([bestWeights, max_sr_ret, max_sr_vol, max_sharpe, wr])

    #'''

    #print('Its location in the array: {}'.format(sharpes.argmax()))

    #print('Optimal Genetic Algorithm Expected Annual Return:
{}'.format((np.exp(max_sr_ret)-1)*252.75))
    #print('Optimal Genetic Algorithm Expected Annual Variance:
{}'.format((np.exp(max_sr_vol)-1)*252.75))
    #print('Optimal Genetic Algorithm sharpe ratio: {}'.format(max_sharpe))

    plt.figure(figsize=(12,8))
    plt.scatter((np.exp(stds.astype(float))-1)*252.75,
(np.exp(means.astype(float))-1)*252.75, c=sharpes, cmap='viridis')
    plt.colorbar(label='Sharpe Ratio')
    plt.xlabel('Volatility')
    plt.ylabel('Return')

    #plt.plot((np.exp(risks)-1)*252.75, (np.exp(returns)-1)*252.75, 'y-o')
    plt.scatter((np.exp(max_sr_vol)-1)*252.75,
(np.exp(max_sr_ret)-1)*252.75,c='red', s=50) # red dot
    #plt.scatter((np.exp(bestStd)-1)*252.75,(np.exp(bestMean)-1)*252.75,c='purple',
s=50) # red dot
    #plt.scatter((np.exp(optimal_risk_free_sharpe[1])-1)*252.75,
(np.exp(optimal_risk_free_sharpe[0])-1)*252.75,c='purple', s=50) # red
dot
    plt.show()

    data0=(Holdout_query).mean(1).cumsum()
    data1=(Holdout_query*weights[sharpes.argmax()]).mean(1).cumsum()

    holdout_returns.append(data0[-1])
    holdout_returns_weighted.append(data1[-1])

    df = pd.concat([data0,data1],axis=1)
    df.columns = ["no weights","weights"]
    df.plot()
    plt.show()
